core/jobs/batch_jobs/manual_voice_artist_names_job.py

```diff
@@ -0,0 +1,504 @@
+# ignore here.
+class FilterCommitLogModelByExplorationIds(beam.DoFn): # type: ignore[misc]
+    """DoFn to filter exploration commit log entry models based on curated
+    exploration IDs.
+    """
+
+    def process(
+        self,
+        exp_commit_log_model: exp_models.ExplorationCommitLogEntryModel,
+        exploration_ids: List[str]
+    ) -> Iterable[exp_models.ExplorationCommitLogEntryModel]:
+        if exp_commit_log_model.exploration_id in exploration_ids:
+            yield exp_commit_log_model
+
+
+# TODO(#15613): Here we use MyPy ignore because the incomplete typing of
+# apache_beam library and absences of stubs in Typeshed, forces MyPy to
+# assume that PTransform class is of type Any. Thus to avoid MyPy's error
+# (Class cannot subclass 'PTransform' (has type 'Any')), we added an
+# ignore here.
+class GetVoiceArtistMetadataModelsFromDict(beam.DoFn): # type: ignore[misc]
+    """DoFn to convert voice artist metadata dicts to their corresponding
+    model instances.
+    """
+
+    total_commit_log_models = 0
+    processed_commit_log_models = 0
+
+    @staticmethod
+    def _convert_to_models(
+        voice_artist_id_to_metadata_mapping: Dict[
+            str, voiceover_models.VoiceoversAndContentsMappingType]
+    ) -> List[voiceover_models.VoiceArtistMetadataModel]:
+        """The method converts the voice artist metadata dicts to their
+        corresponding model instances.
+
+        Args:
+            voice_artist_id_to_metadata_mapping:
+                dict(str, VoiceoversAndContentsMappingType). A dictionary
+                mapping voice artist IDs as keys and nested dicts as values.
+                Each nested dict refers to the voiceovers_and_contents_mapping
+                field in the VoiceArtistMetadataModel.
+
+        Returns:
+            list(VoiceArtistMetadataModel). A list of VoiceArtistMetadataModel
+            instances.
+        """
+
+        voice_artist_metadata_models = []
+
+        with datastore_services.get_ndb_context():
+            for voice_artist_id, metadata_mapping in (
+                    voice_artist_id_to_metadata_mapping.items()):
+                voice_artist_model = (
+                    voiceover_services.
+                    create_voice_artist_metadata_model_instance(
+                        voice_artist_id, metadata_mapping)
+                )
+                voice_artist_metadata_models.append(voice_artist_model)
+
+        return voice_artist_metadata_models
+
+    def process(
+        self,
+        voice_artist_id_to_metadata_mapping: Dict[
+            str, voiceover_models.VoiceoversAndContentsMappingType],
+        exp_commit_log_models: List[exp_models.ExplorationCommitLogEntryModel]
+    ) -> Iterable[voiceover_models.VoiceArtistMetadataModel]:
+        self.total_commit_log_models = len(exp_commit_log_models)
+        self.processed_commit_log_models += 1
+
+        if self.processed_commit_log_models == self.total_commit_log_models:
```

Comment:
I am a bit unsure about this if clause. Do we do similar things elsewhere? Is it resilient/robust to failure? 

E.g. if some commit log model is invalid then we may never reach the equality count -- bit worried about possible cases like that.

Also IIRC some explorations have missing commit log entry models. I suggest dealing with this gracefully. Remember that all we want is some sample voiceovers so it's fine if this doesn't work 100% perfectly to spec, as long as you get that info for each artist. So try to keep the logic as simple as you can.

Reply:
Currently we do not have any validate method for BaseCommitLogEntryModel / ExplorationCommitLogEntryModel. So I am not sure how to confirm whether the model present is valid or not. The assumptions we can do is that if the model is present, then it should at least contain commit changes, user ID, exploration ID, etc using which we can create the voice artist metadata model.

For some explorations if commit log models are not present, then it means we don’t have the data of the user who had committed the changes. This will result in missing voiceovers data (if the missing model is related to voiceover change) and in my opinion it will not be a major issue becuase new manual voiceovers can be added to curated explorations at any point also eventually autogenerated voiceovers will be used.

Example: **Here version 2 is missing**

+----------------------------------------+
Exploration commit model
ID: exploration-exploration_id_2-1
[{'cmd': 'create_new', 'category': 'Algebra', 'title': 'title2'}]
+----------------------------------------+


+----------------------------------------+
Exploration commit model
iD: exploration-exploration_id_2-3
[{'cmd': 'edit_state_property', 'property_name': 'recorded_voiceovers', 'state_name': 'Introduction', 'new_value': {'voiceovers_mapping': {'content_0': {'en': {'filename': 'filename4.mp3', 'file_size_bytes': 3000, 'needs_update': False, 'duration_secs': 20.0}}, 'ca_placeholder_2': {'en': {'filename': 'filename9.mp3', 'file_size_bytes': 3000, 'needs_update': False, 'duration_secs': 20.0}}, 'default_outcome_1': {}}}, 'old_value': {'voiceovers_mapping': {'content_0': {}, 'ca_placeholder_2': {'en': {'filename': 'filename9.mp3', 'file_size_bytes': 3000, 'needs_update': False, 'duration_secs': 20}}, 'default_outcome_1': {}}}}]
+----------------------------------------+

Regarding equality check, I have updated the implementation PTAL.

Reply:
Updated the approach